src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.EnumerateMatches.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Text.RegularExpressions
+{
+    public partial class Regex
+    {
+        /// <summary>
+        /// Searches an input span for all occurrences of a regular expression and returns a <see cref="ValueMatchEnumerator"/> to iterate over the matches.
+        /// </summary>
+        /// <remarks>
+        /// The enumerator returned by this method, as well as the structs returned by the enumerator that wrap each match found in the input are ref structs which
+        /// make this method be amortized allocation free.
```

Comment:
It's probably worth calling out that each match doesn't actually happen until MoveNext is invoked on the enumerator, with one match being performed per MoveNext call.  This is particularly important due to the input being mutable.  strings can't ever change, so it's not observable (other than via the perf impact) as to exactly when the matches are performed with the existing APIs.  But even though `ReadOnlySpan<char>` is itself not writable, it can wrap something that is mutable, and that underlying data source could be change between calls to MoveNext, thereby making it observable exactly when the matching is performed.  (Though obviously we'd highly discourage such use, and maybe even state it's not supported.)